dpar/src/system/trans_system.rs

```diff
@@ -24,3 +25,187 @@ pub trait Transition: Clone + Debug + Eq + Hash + Serialize + DeserializeOwned {
     fn is_possible(&self, state: &ParserState) -> bool;
     fn apply(&self, state: &mut ParserState);
 }
+
+/// Transition lookup table.
+///
+/// Instances of this type provide a transition lookup table. When a fresh
+/// table is created, a transition lookup adds a transition to the table.
+/// If the table is frozen through serialization or the `freeze` method,
+/// the table becomes immutable. Lookups of transitions that are not in the
+/// table will result in a special index (`0`).
+#[derive(Debug, Deserialize, Eq)]
+pub enum TransitionLookup<T>
+where
+    T: Eq + Hash,
+{
+    Fresh(RefCell<Numberer<T>>),
+    Frozen(Numberer<T>),
+}
+
+impl<T> PartialEq for TransitionLookup<T>
+where
+    T: Eq + Hash,
+{
+    fn eq(&self, rhs: &TransitionLookup<T>) -> bool {
+        use self::TransitionLookup::*;
+
+        // Two TransitionLookups are equal if their numberers
+        // are equal.
+        match (self, rhs) {
+            (Frozen(ln), Frozen(rn)) => ln == rn,
+            (Fresh(lrc), Fresh(rrc)) => lrc == rrc,
+            (Frozen(ln), Fresh(rrc)) => ln == &*rrc.borrow(),
+            (Fresh(lrc), Frozen(rn)) => &*lrc.borrow() == rn,
+        }
+    }
+}
+
+impl<T> Serialize for TransitionLookup<T>
+where
+    T: Eq + Hash + Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        use self::TransitionLookup::*;
+
+        match self {
+            Fresh(refcell) => serializer.serialize_newtype_variant(
+                "TransitionLookup",
+                1,
+                "Frozen",
+                &*refcell.borrow(),
+            ),
+            Frozen(ref numberer) => {
+                serializer.serialize_newtype_variant("TransitionLookup", 1, "Frozen", numberer)
+            }
+        }
+    }
+}
+
+impl<T> TransitionLookup<T>
+where
+    T: Clone + Eq + Hash,
+{
+    /// Freeze a transition table.
+    pub fn freeze(self) -> Self {
+        use self::TransitionLookup::*;
+
+        match self {
+            Fresh(cell) => Frozen(cell.into_inner()),
+            frozen => frozen,
+        }
+    }
+
+    /// Length of the transition table.
+    pub fn len(&self) -> usize {
+        use self::TransitionLookup::*;
+
+        match self {
+            Fresh(cell) => cell.borrow().len() + 1,
+            Frozen(numberer) => numberer.len() + 1,
+        }
+    }
+
+    /// Look up a transition.
+    ///
+    /// If the the transition is not in the table, it is added for a
+    /// fresh table. Frozen tables will return the special identifier
+    /// `0` in such cases.
+    pub fn lookup(&self, t: T) -> usize {
```

Comment:
Ownership of t is only required in case of a fresh lookup table where t is added to the numberer. Again, when dealing with Transition Enums probably not really a problem. If we should generalize the struct later on to e.g. number strings we might consider adding to_owned or clone to the trait bounds so that we can pass in a &T and only convert it when necessary.

Reply:
*If we should generalize the struct later on to e.g. number strings we might consider adding to_owned or clone to the trait bounds so that we can pass in a &T and only convert it when necessary.*

I agree that this would be nice. IIRC I decided use moves in `Numberer::add` to avoid  two lookups when the value is absent and `HashMap`s `entry` method requires an owned value.  But since in transition and feature lookups, the lookups succeed in the vast majority of cases, doing an get (with a reference) + insert when absent would be more economical. There was an RFC to make `entry` work with borrowed values, but it seems to be stranded:

https://github.com/rust-lang/rfcs/pull/1769

I'll add this as a todo item to the 'one lookup to rule them all'-issue.

Reply:
We could also consider moving the numberer / lookup into a separate crate. 

For the lemmatizer, I had to implement the same functionality and dealt with similar issues. If there'd be a crate we might spare some people a lot of unnecessary worries and work.